metrics/sources/kubelet/summary.go

```diff
@@ -0,0 +1,367 @@
+	this.decodeCPUStats(containerMetrics, container.CPU, startTime)
+	this.decodeMemoryStats(containerMetrics, container.Memory, startTime)
+	this.decodeFsStats(containerMetrics, RootFsKey, container.Rootfs)
+	this.decodeFsStats(containerMetrics, LogsKey, container.Logs)
+	this.decodeUserDefinedMetrics(containerMetrics, container.UserDefinedMetrics)
+
+	return containerMetrics
+}
+
+func (this *summaryMetricsSource) decodeUptime(metrics *MetricSet, startTime time.Time) {
+	if startTime.IsZero() {
+		return
+	}
+
+	uptime := uint64(time.Since(startTime).Nanoseconds() / time.Millisecond.Nanoseconds())
+	this.addIntMetric(metrics, &MetricUptime, &uptime, startTime)
+}
+
+func (this *summaryMetricsSource) decodeCPUStats(metrics *MetricSet, cpu *stats.CPUStats, startTime time.Time) {
+	if cpu == nil {
+		return
+	}
+
+	this.addIntMetric(metrics, &MetricCpuUsage, cpu.UsageCoreNanoSeconds, startTime)
+
+	// Need to translate from nanocores to millicores.
+	if cpu.UsageNanoCores != nil {
+		const nanoPerMilli uint64 = 1e6
+		cores := *cpu.UsageNanoCores / nanoPerMilli
+		this.addIntMetric(metrics, &MetricCpuUsageRate, &cores, startTime)
+	}
+}
+
+func (this *summaryMetricsSource) decodeMemoryStats(metrics *MetricSet, memory *stats.MemoryStats, startTime time.Time) {
+	if memory == nil {
+		return
+	}
+
+	this.addIntMetric(metrics, &MetricMemoryUsage, memory.UsageBytes, startTime)
+	this.addIntMetric(metrics, &MetricMemoryWorkingSet, memory.WorkingSetBytes, startTime)
+	this.addIntMetric(metrics, &MetricMemoryPageFaults, memory.PageFaults, startTime)
+	this.addIntMetric(metrics, &MetricMemoryMajorPageFaults, memory.MajorPageFaults, startTime)
+}
+
+func (this *summaryMetricsSource) decodeNetworkStats(metrics *MetricSet, network *stats.NetworkStats, startTime time.Time) {
+	if network == nil {
+		return
+	}
+
+	this.addIntMetric(metrics, &MetricNetworkRx, network.RxBytes, startTime)
+	this.addIntMetric(metrics, &MetricNetworkRxErrors, network.RxErrors, startTime)
+	this.addIntMetric(metrics, &MetricNetworkTx, network.TxBytes, startTime)
+	this.addIntMetric(metrics, &MetricNetworkTxErrors, network.TxErrors, startTime)
+}
+
+func (this *summaryMetricsSource) decodeFsStats(metrics *MetricSet, fsKey string, fs *stats.FsStats) {
+	if fs == nil {
+		return
+	}
+
+	fsLabels := map[string]string{LabelResourceID.Key: fsKey}
+	this.addLabeledIntMetric(metrics, &MetricFilesystemUsage, fsLabels, fs.UsedBytes)
+	this.addLabeledIntMetric(metrics, &MetricFilesystemLimit, fsLabels, fs.CapacityBytes)
+	this.addLabeledIntMetric(metrics, &MetricFilesystemAvailable, fsLabels, fs.AvailableBytes)
+}
+
+func (this *summaryMetricsSource) decodeUserDefinedMetrics(metrics *MetricSet, udm []stats.UserDefinedMetric) {
+	for _, metric := range udm {
+		mv := MetricValue{}
+		switch metric.Type {
+		case stats.MetricGauge:
+			mv.MetricType = MetricGauge
+		case stats.MetricCumulative:
+			mv.MetricType = MetricCumulative
+		case stats.MetricDelta:
```

Comment:
Add delta handling in pod aggragtor. We should not aggregate delta metrics as we don't aggregate cumulatives.